### PR TITLE
[2/Y] Add a POST method for comments and display comments.

### PR DESCRIPTION
**Pull request by nikhitashanker**
Currently, the elements in the comments data structure are hard-coded and the display of the comments is in JSON format. In this change, we add a form element to the portfolio page. We also add a doPost() function to the DataServlet.java file that takes data from the POST requests and populates the comments data structure. Additionally, the display of comments is changed to text format.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.google.sps.servlets;
+
+/** Object that stores comment data. */
+public class Comment {
+  String text;
+  String commenterName;
+  String commenterEmail;
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/Comment.java:7]: Try to follow alphabetic order to declare variables.

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -105,4 +106,27 @@ function changeNavbarStickiness() {
...
+ * the comments.
+ */
+function getComments() {
+  fetch("/data").then(response => response.json()).then(comments => {
+    const commentContainer = document.getElementById("comment-container");
+
+    // Build display of comments. 
+    comments.forEach((comment) => {
+      commentContainer.appendChild(createHeadingElement(comment.commenterName + "(" 
+          + comment.commenterEmail + ") says" + "\"" + comment.text +"\""));
```
Comment (liangqigong) [inline portfolio/src/main/webapp/script.js:122]: Can you put 
comment.commenterName + "(" + comment.commenterEmail + ") says" + "\"" + comment.text in a public function in Comment, and make all these 3 variables private?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -14,19 +14,68 @@
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get input from the form.
+    String text = getParameter(request, "comment-text", "");
+    String commenterName = getParameter(request, "commenter-name", "Anonymous");
+    String commenterEmail = getParameter(request, "commenter-email", "Unknown");
+
+    // Create new Comment object to store input.
+    Comment newComment = new Comment(text, commenterName, commenterEmail);
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:53]: Usually I would like to inline variables if it is only used once, but I understand that this line could become very complicate so it depends on you.

But I think you can definitely inline newComment.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -14,19 +14,68 @@
...
+   * Converts List of comments into a JSON using the gson library.
+   */
+  private String convertToJson(List<Comment> comments){
+    Gson gson = new Gson();
+    return gson.toJson(comments);
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:72]: The format of comment is very strange. Did you use any auto formatter?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -14,19 +14,68 @@
...
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null || value.isEmpty()) {
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:76]: return (value == null || value.isEmpty()) ? defaultValue : value;

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -14,19 +14,68 @@
...
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comments data. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private List<Comment> comments = new ArrayList<Comment>();
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("text/html;");
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:35]: Remove?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -14,19 +14,68 @@
...
-    response.getWriter().println("<h1>Hello world!</h1>");
+
+    // Convert the comments into JSON.
+    String json = convertToJson(comments);
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:38]: I think you can inline this.

BTW, Variables are typically declared just before they are first used (to minimize scope).
go/java-style#s4.8.2.2-variables-limited-scope
go/ej3e-57

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -14,19 +14,68 @@
...
+
+    // Redirect to same HTML page.
+    response.sendRedirect("/index.html");
+  }
+
+  /*
+   * Converts List of comments into a JSON using the gson library.
+   */
+  private String convertToJson(List<Comment> comments){
+    Gson gson = new Gson();
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:66]: inline.

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (nikhitashanker) [reply portfolio/src/main/webapp/script.js:122]: I have made the three variables private and added getter functions. From my understanding, a public function that returns comment.commenterName + "(" + comment.commenterEmail + ") says" + """ + comment.text in Comment may not be helpful since the javascript file is using the response which is in JSON format.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:72]: Done -  removed the extra asterisk at the top. Yes, prettier did not flag this. I tried to follow the example here since this is javadoc. https://google.github.io/styleguide/javaguide.html#s7-javadoc

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:53]: Done - will try to inline variables only used once.

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java (hunk 1 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/Comment.java:7]: Done.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:38]: Done. I see, thanks for the references - I have inlined this.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:35]: Done.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:76]: Done - use of a ternary operator.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done.

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (hlz03) [inline portfolio/src/main/webapp/script.js:122]: You can use string template 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

Comment (nikhitashanker) [reply portfolio/src/main/webapp/script.js]: Done. I added a helper function that uses string template.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (liangqigong) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:63]: Ah I see.

File: .htmlvalidate.json [hunk 9]
```diff
@@ -6,6 +6,7 @@
     "html-validate:recommended"
   ],
   "rules": {
-    "void-style": "off"
+    "void-style": "off",
+    "prefer-button": "warn"
```
Comment (liangqigong) [inline .htmlvalidate.json:10]: Let's put this in a different PR?

Comment (liangqigong) [reply .htmlvalidate.json:10]: Ah it is merged. NVM

Comment (nikhitashanker) [reply .htmlvalidate.json:10]: Noted, will do next time. I was trying to make sure I didn't push something with a failed build.
